File: iotdb/src/main/java/org/apache/iotdb/db/utils/QueryUtils.java
/**
 * Copyright © 2019 Apache IoTDB(incubating) (dev@iotdb.apache.org)
 * <p>
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.iotdb.db.utils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import org.apache.iotdb.db.engine.modification.Deletion;
import org.apache.iotdb.db.engine.modification.Modification;
import org.apache.iotdb.db.engine.modification.ModificationFile;
import org.apache.iotdb.tsfile.file.metadata.ChunkMetaData;

public class QueryUtils {

  /**
   * modifyChunkMetaData iterates the chunkMetaData and applies all available modifications on it to
   * generate a ModifiedChunkMetadata.
   * @param chunkMetaData the original chunkMetaData.
   * @param modifications all possible modifications.
   * @return a list of all generated ModifiedChunkMetadata.
   */
  public static void modifyChunkMetaData(List<ChunkMetaData> chunkMetaData,
                                         List<Modification> modifications) {
    int modIndex = 0;

    for (int metaIndex = 0; metaIndex < chunkMetaData.size(); metaIndex++) {
      ChunkMetaData metaData = chunkMetaData.get(metaIndex);
      for (int j = modIndex; j < modifications.size(); j++) {
        // iterate each modification to find the max deletion time
        Modification modification = modifications.get(modIndex);
        if (modification.getVersionNum() > metaData.getVersion()) {
          if (modification instanceof Deletion) {
            Deletion deletion = (Deletion) modification;
            if (metaData.getDeletedAt() < deletion.getTimestamp()) {
              metaData.setDeletedAt(deletion.getTimestamp());
              modIndex = j;
            }
          }
        } else {
          // skip old modifications for next metadata
          modIndex++;
        }
      }
    }
    // remove chunks that are completely deleted
    chunkMetaData.removeIf(metaData -> metaData.getDeletedAt() >= metaData.getEndTime());
  }
}

File: tsfile-go/src/tsfile/common/utils/QueryUtils.go
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package utils

import (
	"sort"
	"strings"
)

// MergeStrings sorts and merges two string lists in ascent order.
func MergeStrings(strsA []string, strsB []string) []string {
	if strsB == nil {
		return strsA
	} else if strsA == nil {
		return strsB
	}

	lenA, lenB := len(strsA), len(strsB)
	indexA, indexB := 0, 0
	sort.Strings(strsA)
	sort.Strings(strsB)
	var ret []string
	for indexA < lenA && indexB < lenB {
		order := strings.Compare(strsA[indexA], strsB[indexB])
		if order == 0 {
			ret = append(ret, strsA[indexA])
			indexA++
			indexB++
		} else if order < 0 {
			ret = append(ret, strsA[indexA])
			indexA++
		} else {
			ret = append(ret, strsB[indexB])
			indexB++
		}
	}
	for indexA < lenA {
		ret = append(ret, strsA[indexA])
		indexA++
	}
	for indexB < lenB {
		ret = append(ret, strsB[indexB])
		indexB++
	}
	return ret
}

// TestCommonStrs tests if two string slices have common strings using sort and merge.
func TestCommonStrs(strsA []string, strsB []string) bool {
	if strsB == nil || strsA == nil {
		return false
	}
	lenA, lenB := len(strsA), len(strsB)
	indexA, indexB := 0, 0
	sort.Strings(strsA)
	sort.Strings(strsB)
	for indexA < lenA && indexB < lenB {
		order := strings.Compare(strsA[indexA], strsB[indexB])
		if order == 0 {
			return true
		} else if order < 0 {
			indexA++
		} else {
			indexB++
		}
	}
	return false
}
